pc2ccs/pc2v9
# I122 multiple team logins

### Description of what the PR does
Provides support for multiple simultaneous logins for a single team (although not for any other client type).  Adds an Admin "Allow multiple team logins" checkbox (defaults to _unchecked_); when checked then subsequent duplicate team logins do not generate Security Violations and do not force-logoff the existing team login(s).  Works with both the Fat Team Client and WTI.

### Issue which the PR fixes
Fixes #122 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 10, Eclipse 2019-12, Java 8_201.

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

Note:  the following tests are expected to be done in sequence; later tests assume their start-state is the the state at the end of the preceding test.

#### Test 1: Multiple Logins Default To Disabled
- Start a server (call it "Site 1") with a contest configured with multiple team accounts, at least one Judge account, and at least one problem.
- Login an Admin, start the contest.
- Login as a Judge.  (If you want to automate Submission Testing/Judgement, below, then configure the Judge to Auto-Judge a problem; alternatively, you can Judge submissions manually).
- Start a Fat Team Client logged in as Team 1.
- Start a _second_ Fat Team Client and try to login _again_ as Team 1.
- Verify that the second login generates a Security Log screen and forces off the first Team 1 login (because the default is "do NOT allow multiple logins").

#### Test 2: Multiple Logins Work When Enabled
- Go to the Admin **ConfigureContest>Settings** tab, scroll down to the "Team Settings" area, and check the "Allow multiple team logins" checkbox.
- Start another Fat Team Client and login _again_ as Team 1.
- Verify that there are NOT new Security Log entries about "Duplicate Logins" or "Forced Logoffs", and that a second instance of the Fat Team Client is now logged in.

#### Test 3: Forced Logoff Works
- Go to the Admin **Run Contest>Logins** tab; verify that both Team 1 logins are displayed.
- Go to the Server **Logins** tab; verify that both Team 1 logins are displayed.
- Select one of the Team 1 logins (on either the Server or the Admin) and click "Logoff".
- Verify that the selected login entry disappears from both the Server and the Admin **Logins** tabs.
- Verify that the Fat Team Client GUI associated with the selected (forced-off) login does a "Countdown" and then exits.

#### Test 4: New Login After Disabling Allow-Multiple-Logins Forces Off Duplicates
- Start a new Fat Team Client and login again as Team 1.
- Verify that there are now (again) two Team 1 clients logged in, with two separate Fat Team Client GUIs.
- Go to the Admin **ConfigureContest>Settings** tab and _uncheck_ the "Allow multiple team logins" checkbox.
- Start _another_ Fat Team Client and login again as Team 1.
- Verify that the new Team 1 login forces off the previous Team 1 logins: all other Team 1 login GUIs should be removed and a Security Log entry should indicate the duplicate/forced-off logins.

#### Test 5: WTI Works With Multiple Logins
- Unzip the latest build of the WTI-API project and go into the WebTeamInterface-1.1 folder.
- If you are going to run the WTI on a different machine from the one on which the PC2 server is running, edit the **pc2v9.ini** file in the WTI folder to point to the PC2 Server's IP.
- Start the WTI server using the command **./bin/pc2wti**.
- Open a browser window and point it to the IP:port of the WTI server (default is **localhost:8080**).
- Verify that the "WTI Login Screen" is displayed.
- Login as Team 1.
- Verify that the WTI displays the "main team screen", that the PC2 Server and Admin **Login** tabs show multiple Team 1 logins (including the new "WTI login"), and that the previous "Fat Team Client" login for Team 1 is _still present_.

#### Test 6: Submissions Work For Multiple Logins
- Start another Fat Team Client and login as Team 1.
- Start another browser window, connect to the WTI server, and login as Team 1.
(You should now have TWO Fat Team Client GUIs plus TWO WTI login clients; that is, there should be _four_ logins for Team 1: two with the Fat Client and two with WTI.)
- Verify that there are four logins for Team 1 on both the Server and Admin **Logins** tabs.
- On one of the Fat Team Client GUIs, select a problem, a language, and a main file for that problem and language.
- Click **Submit** on the Fat Team Client, and confirm the submission.
- Verify that ALL logged-in Team 1 clients -- Fat Clients and WTI -- get a "Submission Received" notification.
(Note: the "Submission Received" notification on the WTI is not a permanent pop-up like it is on the Fat Team client; it is a "Snackbar" pop-up that displays at the bottom of the screen and goes away after a short delay.)
- If the Judge wasn't configured for AutoJudging (Test 1, above), go to the Judge and manually judge the submission.
- Verify that ALL logged-in Team 1 clients (Fat and WTI) got a notification of the Submission Judgement.
- On one of the WTI browser clients, click **Submit Problem**.
- Select a problem, a language, and a main file for that problem and language, then click **Submit*.
- Verify that ALL logged-in Team 1 clients -- Fat Clients and WTI -- get a "Submission Received" notification.
- If the Judge wasn't configured for AutoJudging (Test 1, above), go to the Judge and manually judge the submission.
- Verify that ALL logged-in Team 1 clients (Fat and WTI) got a notification of the Submission Judgement.

#### Test 7: Works For Remote Sites
- Start a second PC2 Server (say, "Site 2"). Make sure that the "**remoteServer**" entry in Site 2's **pc2v9.ini** points to the IP of Site 1.
- Start multiple Fat Team Clients whose **pc2v9.ini** points them at Site 2's server.
- Start a WTI server with a **pc2v9.ini** pointing to Site 2's server; call this the "Site 2 WTI server".
- Start multiple team browser sessions pointing to the URL for the Site 2 WTI Server.
- Verify that multiple Fat Team Clients and multiple WTI clients can login when connected to Site 2.
- Submit from the Site 2 team clients; verify that "submission received" notifications are received by all team clients logged in as the same team.
- Judge the submitted run on the Site 1 judge (or using AutoJudging on Site 1).
- Verify that all team clients (Fat and WTI) receive the proper "Run Judgement" notifications.

If all the above works, the PR works correctly: multiple team logins are now supported.
